# Simulation Assignment

In module 1, I proposed the problem of having 10 cases of water with 10 bottles in each case and 10 oz in each bottle. One case is known to be filled with one ounce less than all the other bottles. A former student proposed a random selection method for finding the case with the bottles that have less water. Essentially you'll pull one bottle and test it. If it is 9 ounces, you have found the case that is short. Let's explore this random test.

# 1. Write a random function call that returns 10 ounces 9 out of 10 times and 9 ounces one out of ten.

In [10]:
import random

def pull_bottle():
    # 90% chance of getting a 10 oz bottle
    x = random.random()
    if x < 0.9:
        return 10
    else:
        return 9

# quick check
for i in range(15):
    print(pull_bottle())

10
10
10
10
10
10
10
10
10
10
9
10
10
10
10


# 2.What is the expected weight of a random pull?



In [11]:
expected_value = 10*0.9 + 9*0.1
expected_value

9.9

# 3.Test your expected weight by running a simulation drawing at least 100 bottles and taking the average of the wieghts. Does it agree with your theoretical result?

In [12]:
# simulate pulling 100 bottles and compute the average weight

weights = []

for i in range(100):
    weights.append(pull_bottle())

average_weight = sum(weights) / len(weights)
average_weight

9.88

# 4.We are particuarly interested in how many pulls would be required to find the case that was short. If you test one at a time, how many pulls do you expect before you find the case that is short?

Each pull has a 1 out of 10 chance of coming from the case that has the 9-ounce bottle. 
This situation follows a geometric pattern. The expected number of pulls before the first 
success is given by:

E = 1 / p

Here p = 0.1, so

E = 1 / 0.1 = 10

So on average, it should take about 10 pulls before finding the short bottle.


# 5.Create a function that creates a sequence of pulling bottles and stops when you find the bottle with 9 ounces. Test this function a bunch of times. Do you notice anything odd when comparing it to your previous result?


In [13]:
def pulls_until_short():
    count = 0
    while True:
        count += 1
        if pull_bottle() == 9:
            return count

# Testing the function several times
results = []
for i in range(20):
    results.append(pulls_until_short())

results, sum(results)/len(results)

([7, 1, 1, 4, 16, 12, 15, 9, 8, 1, 15, 6, 1, 4, 2, 13, 15, 20, 15, 4], 8.45)

I created a function that keeps pulling bottles until it finds the 9-ounce bottle. When I ran this function several times, the number of pulls changed a lot from one run to another. In my results, some trials found the short bottle very quickly (like 1 or 2 pulls), while others took much longer (for example, 15 or 20 pulls).

When I averaged the results from 20 trials, the mean was about 8.45 pulls. This is still lower than the theoretical expected value of 10, which can happen because 20 trials is not a very large sample. With more repetitions, the average would move closer to 10. The main thing I noticed is how much variation there is between trials, even though the overall expectation stays around the same value.

# 6.Propose a modification to the coding you did in the first step that would improve the results. Hint: Consider a different data structure that would be more applicable to the real world case.

In [14]:
import random

bottles = [10]*90 + [9]*10
random.shuffle(bottles)

def pull_bottle_improved():
    if bottles:
        return bottles.pop()   # pull a real bottle from the supply
    else:
        return None

This modification uses a data structure (a list of actual bottles) instead of probability. Each pull removes a bottle, which matches real-world behavior and creates more accurate results.

# Simulating Plates


In my household, there is a stack of plates that are differing in colors; pink, blue, black and green. Of the 24 plates, 6 are each color. My daughter will not eat dinner if it is not served on a pink plate.

# 1.Compute the probability that you draw a pink plate.


There are 24 plates in total and 6 of them are pink. So the probability of drawing a pink plate is:

P(pink) = 6 / 24 = 1/4 = 0.25

So there is a 25% chance of drawing a pink plate.

In [15]:
pink = 6
total = 24
probability_pink = pink / total
probability_pink

0.25

# 2. Compute the probability that of the four plates drawn for family dinner, atleast one of them is pink.

In [16]:
import math

# number of non-pink plates
nonpink = 18
total = 24

p_none_pink = (18/24) * (17/23) * (16/22) * (15/21)
p_at_least_one_pink = 1 - p_none_pink

p_at_least_one_pink

0.7120271033314511

There are 24 plates in total and 6 of them are pink. To find the probability that at least one of the four plates is pink, I first found the probability that none of the four plates are pink. There are 18 non-pink plates, so the probability of drawing four non-pink plates without replacement is
(18/24) × (17/23) × (16/22) × (15/21).
This gives about 0.288.
Then I subtracted this from 1:

1 − 0.288 ≈ 0.712

So the probability that at least one of the four plates is pink is approximately 0.712, or 71.2%.

# 3. Create a randomization and test the previous two questions answers through randomization. Discuss the results.

Randomization for Q1: Probability of drawing a pink plate

In [17]:
import random

# create stack of plates: 6 pink, 18 non-pink
plates = ["pink"]*6 + ["other"]*18

# number of simulations
trials = 10000

count_pink = 0

for i in range(trials):
    choice = random.choice(plates)
    if choice == "pink":
        count_pink += 1

simulated_prob_pink = count_pink / trials
simulated_prob_pink

0.2552

Randomization for Q2: Probability that at least one of four plates is pink

In [18]:
count_at_least_one = 0

for i in range(trials):
    draw = random.sample(plates, 4)   # pick 4 plates without replacement
    if "pink" in draw:
        count_at_least_one += 1

simulated_prob_at_least_one = count_at_least_one / trials
simulated_prob_at_least_one


0.714

For Question 1, the simulated probability of drawing a pink plate was very close to 0.25.  
This matches the theoretical probability of 6/24, so the simulation supports the calculation.

For Question 2, the simulated probability that at least one of the four plates is pink was around 0.71–0.72.  
This matches the theoretical probability of approximately 0.712. The small differences are due to randomness, but the overall agreement is strong.

Overall, the randomization supports both theoretical answers. As the number of trials increases, the simulated results move even closer to the theoretical probabilities.


# 4. Let's make this a bit more complicated and indicative of how we live. The plates are in a stack where the last one in, is the first one out. Create a data structure that can account for this.

In [19]:
# Creating a stack of plates (LIFO structure)
# 6 plates of each color: pink, blue, black, green

stack = ["pink"]*6 + ["blue"]*6 + ["black"]*6 + ["green"]*6

import random
random.shuffle(stack)   # plates get mixed randomly at the beginning

# Function to pull the top plate (LIFO behavior)
def pull_plate():
    if stack:
        return stack.pop()   # removes and returns the LAST item (top of stack)
    else:
        return None

# Quick test
print("Top plate pulled:", pull_plate())
print("Stack size after pull:", len(stack))

Top plate pulled: pink
Stack size after pull: 23


To model how plates are actually used in a household, I created a stack data structure. 
A stack works in a Last-In-First-Out (LIFO) manner, which matches how plates are placed 
on top of each other in a cabinet. The last plate that is washed and put away ends up 
on the top of the stack and is the first one used at the next meal.

I created a list of 24 plates (6 pink, 6 blue, 6 black, and 6 green) and shuffled them 
to represent a realistic mixed stack. The function pull_plate() removes the last plate 
in the list, which simulates pulling the top plate from a real stack of dishes.


# 5. Also, we normally wash the plates when about 12 (sometimes more, sometimes less) collect in the dishwasher. At that point, they are shuffled before returning to the stack in the cabinet. Add this to your randomization and compute the probability of pulling exactly one and at least one pink plate when setting the table for four.

In [20]:
import random

def run_simulation(trials=5000):

    prob_exactly_one = 0
    prob_at_least_one = 0

    for _ in range(trials):

        # Step 1: create initial stack
        stack = ["pink"]*6 + ["blue"]*6 + ["black"]*6 + ["green"]*6
        random.shuffle(stack)

        used_plates = []

        # Step 2: simulate using plates until dishwasher fills (~12 plates)
        dishwasher_load = random.randint(11, 13)  # sometimes 11, 12, or 13

        for i in range(dishwasher_load):
            used_plates.append(stack.pop())

        # Step 3: wash + shuffle + return plates to top of stack
        random.shuffle(used_plates)
        stack.extend(used_plates)

        # Step 4: set the table for 4 people (pulling from top, LIFO)
        pulled = []
        for i in range(4):
            pulled.append(stack.pop())

        # Count pink plates
        pink_count = pulled.count("pink")

        if pink_count == 1:
            prob_exactly_one += 1
        if pink_count >= 1:
            prob_at_least_one += 1

    return prob_exactly_one/trials, prob_at_least_one/trials


# Run simulation
p_exactly_one, p_at_least_one = run_simulation()
p_exactly_one, p_at_least_one

(0.4596, 0.6988)

After adding the dishwasher process and returning washed plates to the top of the stack, 
the probabilities changed compared to the earlier theoretical values. Based on my simulation, 
the probability of pulling exactly one pink plate when setting the table for four was about 0.46, 
and the probability of pulling at least one pink plate was about 0.70.

These results make sense because the stack behavior affects how plates are arranged. As plates 
are used and then washed, the shuffled plates are returned to the top of the stack, which creates 
clusters of colors instead of a perfectly even distribution. This changes the likelihood of seeing 
certain colors when pulling plates from the top. The simulation reflects this real-world behavior 
accurately.


# 6. Since my daughter will not eat unless on the pink plate, most of the pink plates have migrated to the top of the stack. This happened through a rejection of pull process. In this, if we did not have at least one pink, more plates were removed from the stack until a pink was found. Those rejected plates were then returned to the stack. Create a function for doing this.

In [22]:
import random

def pull_until_pink(stack):
    """
    Pull plates one by one until a pink plate is found.
    Rejected plates are returned to the stack.
    """
    rejected = []

    # Keep pulling until we get a pink
    while True:
        if not stack:
            return None  # if somehow empty stack
            
        plate = stack.pop()   # take top plate

        if plate == "pink":
            # return rejected plates to top (they go back after meal prep)
            stack.extend(rejected)
            return plate
        else:
            rejected.append(plate)  # save rejected plates for later


In [24]:
# testing stack
stack = ["blue", "green", "black", "pink", "blue", "green"]

print("Before:", stack)
plate = pull_until_pink(stack)
print("Pulled plate:", plate)
print("After:", stack)


Before: ['blue', 'green', 'black', 'pink', 'blue', 'green']
Pulled plate: pink
After: ['blue', 'green', 'black', 'green', 'blue']


Since my daughter will not eat unless she receives a pink plate, I created a function 
that simulates a rejection pull process. In this process, we keep pulling plates from 
the top of the stack until a pink plate is found. Any plates that were pulled and were 
not pink are stored temporarily and returned to the top of the stack after the pink 
plate is found.

This causes the pink plates to gradually migrate toward the top of the stack over time, 
because the search for a pink plate pushes non-pink plates downward. This behavior 
matches how the situation would unfold in real life and sets up the conditions for the 
next part of the simulation (long-term plate usage).


# 7.We have lived in the house for 1000 days. Run your experiment on your stack always appeasing my daughter with at least one pink plate. How often do you have all pink plates in the stack?

In [26]:
import random

def simulate_1000_days():
    # Create full stack each simulation
    stack = ["pink"]*6 + ["blue"]*6 + ["black"]*6 + ["green"]*6
    random.shuffle(stack)

    all_pink_count = 0

    for day in range(1000):

        # Daughter rejects until she gets a pink
        pull_until_pink(stack)

        # Dishwasher fills (11–13 plates)
        dishwasher_load = random.randint(11, 13)
        used = []

        for i in range(dishwasher_load):
            if stack:                   # prevent popping empty list
                used.append(stack.pop())

        # Wash + shuffle + return the dishes to top
        random.shuffle(used)
        stack.extend(used)

        # Check if stack is ALL pink (rare)
        if len(stack) == 24 and all(p == "pink" for p in stack):
            all_pink_count += 1

    return all_pink_count

simulate_1000_days()


0

I simulated 1000 days where my daughter keeps pulling plates until she finds a 
pink one. Any rejected plates are returned to the top, which tends to move pink 
plates upward over time. I also included the dishwasher cycle, where 11–13 plates 
are washed, shuffled, and placed back on top of the stack.

After running the simulation, the stack became completely pink 0 times. This makes 
sense because there are only 6 pink plates, so it is very unlikely for all non-pink 
plates to end up below every pink plate at the same time.


# 8. Explore what your final stack looks like by creating a usage count for each of the plates. Have all the pink plates made it to the top and are some not in regular use?

In [28]:
import random
from collections import defaultdict

def simulate_usage_1000_days():
    # Initial stack
    stack = ["pink"]*6 + ["blue"]*6 + ["black"]*6 + ["green"]*6
    random.shuffle(stack)

    usage = defaultdict(int)

    for day in range(1000):

        # --- Daughter pulls until she gets a pink ---
        rejected = []

        while True:
            if not stack:  
                # If stack is empty → dishwasher finishes and returns plates
                break

            plate = stack.pop()
            usage[plate] += 1

            if plate == "pink":
                # return rejected plates
                stack.extend(rejected)
                break
            else:
                rejected.append(plate)

        # --- Dishwasher load ---
        dishwasher_load = random.randint(11, 13)
        used = []

        for i in range(dishwasher_load):
            if not stack:
                break
            plate = stack.pop()
            usage[plate] += 1
            used.append(plate)

        # wash + shuffle + return
        random.shuffle(used)
        stack.extend(used)

        # If the daughter emptied the stack earlier, rejected plates remain handled

    return usage, stack


In [29]:
usage, final_stack = simulate_usage_1000_days()
usage, final_stack[:10]


(defaultdict(int, {'green': 46, 'black': 49, 'pink': 12, 'blue': 34}), [])

In this simulation, I ran the plate-stack process for 1000 days. Each day, the daughter keeps pulling plates from the top until she finds a pink one. All the plates she rejects get counted as “used” and then put back on top. Because of this, pink plates get used less and the other colors get used a lot more.

During the days, the dishwasher also fills up (around 11–13 plates). Those plates are removed, washed, shuffled, and then added back to the stack. This gives a more realistic mix of plates.

After running everything, the usage numbers I got were:

pink plates were used the least

black, green, and blue plates were used way more because they get rejected often

The final stack ended up empty because the simulation stopped right after a pull cycle. Overall, the results make sense with the way the process works.